### DISCUSSION TITLE
In rare cases only abbreviated Stockfish output

### DISCUSSION THREAD

**Discussion by Mercutio1243** [official-stockfish/Stockfish#5099] [Q&A]
I am using Stockfish to identify valuation and best continuation (PV) for a position. For this, I use the commands
```
position fen FEN
go movetime 150
```

Normally, this works very fine with output in the following format:
```
info string NNUE evaluation using nn-ad9b42354671.nnue enabled
info depth 1 seldepth 1 multipv 1 score cp 18 nodes 20 nps 6666 hashfull 0 tbhits 0 time 3 pv e2e4
info depth 2 seldepth 2 multipv 1 score cp 46 nodes 66 nps 22000 hashfull 0 tbhits 0 time 3 pv d2d4
info depth 3 seldepth 2 multipv 1 score cp 51 nodes 120 nps 40000 hashfull 0 tbhits 0 time 3 pv e2e4
info depth 4 seldepth 2 multipv 1 score cp 58 nodes 144 nps 48000 hashfull 0 tbhits 0 time 3 pv d2d4
info depth 5 seldepth 2 multipv 1 score cp 58 nodes 174 nps 58000 hashfull 0 tbhits 0 time 3 pv d2d4 a7a6
info depth 6 seldepth 7 multipv 1 score cp 34 nodes 1303 nps 260600 hashfull 0 tbhits 0 time 5 pv e2e4 c7c5 g1f3 b8c6 c2c3
info depth 7 seldepth 6 multipv 1 score cp 29 nodes 3126 nps 347333 hashfull 1 tbhits 0 time 9 pv d2d4 g8f6 e2e3 d7d5 c2c4 d5c4
info depth 8 seldepth 7 multipv 1 score cp 26 nodes 5791 nps 445461 hashfull 4 tbhits 0 time 13 pv g1f3 g8f6 d2d4 d7d5 e2e3
info depth 9 seldepth 9 multipv 1 score cp 31 nodes 8541 nps 474500 hashfull 5 tbhits 0 time 18 pv g1f3 c7c5 e2e4 e7e6 d2d4 c5d4 f3d4
info depth 10 seldepth 13 multipv 1 score cp 25 nodes 20978 nps 537897 hashfull 10 tbhits 0 time 39 pv e2e4 c7c5 g1f3 b8c6 f1c4 e7e6 e1g1 g8f6
info depth 11 seldepth 13 multipv 1 score cp 32 nodes 29040 nps 537777 hashfull 14 tbhits 0 time 54 pv e2e4 c7c5 c2c3 g8f6 e4e5 f6d5 d2d4
info depth 12 seldepth 14 multipv 1 score cp 38 nodes 41207 nps 564479 hashfull 18 tbhits 0 time 73 pv e2e4 e7e6 d2d4 d7d5 b1c3 d5e4 c3e4
info depth 13 seldepth 15 multipv 1 score cp 31 nodes 60308 nps 579884 hashfull 25 tbhits 0 time 104 pv e2e4 c7c5 g1f3 d7d6 d2d4 c5d4 f3d4 b8c6 c1e3 g8f6 b1c3 g7g6
info depth 14 seldepth 18 multipv 1 score cp 35 nodes 97789 nps 578633 hashfull 44 tbhits 0 time 169 pv d2d4 g8f6 c2c4 e7e6 g1f3 d7d5 g2g3 f8e7 b1c3 c7c5
info depth 15 seldepth 21 multipv 1 score cp 31 nodes 142447 nps 560814 hashfull 64 tbhits 0 time 254 pv d2d4 g8f6 c2c4 e7e6 g1f3 d7d5 c1g5 f8e7 b1c3 e8g8 e2e3 h7h6 g5h4 f6e4 h4e7 d8e7
info depth 16 seldepth 21 multipv 1 score cp 40 nodes 199887 nps 552174 hashfull 88 tbhits 0 time 362 pv d2d4 d7d5 c2c4 d5c4 g1f3 a7a6 b1c3 c7c6 a2a4
info depth 17 seldepth 19 multipv 1 score cp 32 nodes 268004 nps 552585 hashfull 120 tbhits 0 time 485 pv d2d4 g8f6 c2c4 e7e6 g2g3 d7d5 g1f3 f8e7 f1g2 e8g8 e1g1 c7c5 c4d5 e6d5 b1c3 h7h6 d4c5 e7c5
info depth 18 seldepth 19 multipv 1 score cp 32 nodes 277075 nps 553043 hashfull 122 tbhits 0 time 501 pv d2d4 g8f6 c2c4 e7e6 g2g3 d7d5 g1f3 f8e7 f1g2 e8g8 e1g1 c7c5 c4d5 e6d5 b1c3 h7h6 d4c5 e7c5
bestmove d2d4 ponder g8f6
```

I use the second to last line to automatically extract the information needed. However, in very rare cases (maybe 1 in 10.000?), Stockfish provides the output in an unexpected format as follows:

```
info string NNUE evaluation using nn-ad9b42354671.nnue enabled 
bestmove b2b3
```

This seems to happen quite randomly.  

**Can anybody please explain, what would trigger Stockfish to provide a result like the one above and how I might avoid it?**

With the approach, I was relying on the UCI documentation, which states that the info part should always be included: 

> Directly before that the engine should send a final "info" command with the final search information, the the GUI has the complete statistics about the last search.

Ps: I am using Stockfish 15.1

**Comment by PGG106** [comment]
a fen of a position where you get a lone "bestmove" output would definitely be helpful

**Reply by Mercutio1243** [reply]
@PGG106 Yes, definitely, but I am afraid, I cannot provide this as of now: I have a large set of fens where it has happened for someone else but when running through the fens myself, the error did not occur.

I will set something up to identify a specific fen going forward. In the meantime, if anybody has any ideas, I would still be super grateful!

**Reply by PGG106** [reply]
are you using books perchance? there are sf forks that allow polyglot books for example

**Reply by Mercutio1243** [reply]
I am using Stockfish 15.1 as is, without manually adding any opening book database - if this is what you are referring to?

**Reply by Mercutio1243** [reply]
In general, I am correct in the assumption that this output should generally not occur?

**Reply by PGG106** [reply]
yes, the output is weird (but not impossible), but it's hard to tell what is happening without being able to reproduce it, why are you using an sf version that is so old?

**Comment by bftjoe** [comment]
Are you really using only 150 milliseconds?

**Reply by Mercutio1243** [reply]
In this specific occasion yes, this is for a first assessment, and I am running through a lot of positions. Normally, with 150 ms, I get to a depth of ~14 and seldepth of ~20, which seems to be sufficient for my use case.

**Reply by bftjoe** [reply]
You should not be surprised in the slightest if output is abbreviated at such short TC. Use a longer one.

**Reply by PGG106** [reply]
the best guess is it somehow can't even complete a depth 1 search, therefore it has no info to show, but that would be more or less reproducible given a fen where this happens

**Reply by Mercutio1243** [reply]
In 9999 out of 10000 cases I reliably get output comparable to the following:
 
```
info string NNUE evaluation using nn-ad9b42354671.nnue enabled
info depth 1 seldepth 1 multipv 1 score cp 18 nodes 20 nps 6666 hashfull 0 tbhits 0 time 3 pv e2e4
info depth 2 seldepth 2 multipv 1 score cp 46 nodes 66 nps 22000 hashfull 0 tbhits 0 time 3 pv d2d4
info depth 3 seldepth 2 multipv 1 score cp 51 nodes 120 nps 40000 hashfull 0 tbhits 0 time 3 pv e2e4
info depth 4 seldepth 2 multipv 1 score cp 58 nodes 144 nps 48000 hashfull 0 tbhits 0 time 3 pv d2d4
info depth 5 seldepth 2 multipv 1 score cp 58 nodes 174 nps 58000 hashfull 0 tbhits 0 time 3 pv d2d4 a7a6
info depth 6 seldepth 7 multipv 1 score cp 34 nodes 1303 nps 260600 hashfull 0 tbhits 0 time 5 pv e2e4 c7c5 g1f3 b8c6 c2c3
info depth 7 seldepth 6 multipv 1 score cp 29 nodes 3126 nps 347333 hashfull 1 tbhits 0 time 9 pv d2d4 g8f6 e2e3 d7d5 c2c4 d5c4
info depth 8 seldepth 7 multipv 1 score cp 26 nodes 5791 nps 445461 hashfull 4 tbhits 0 time 13 pv g1f3 g8f6 d2d4 d7d5 e2e3
info depth 9 seldepth 9 multipv 1 score cp 31 nodes 8541 nps 474500 hashfull 5 tbhits 0 time 18 pv g1f3 c7c5 e2e4 e7e6 d2d4 c5d4 f3d4
info depth 10 seldepth 13 multipv 1 score cp 25 nodes 20978 nps 537897 hashfull 10 tbhits 0 time 39 pv e2e4 c7c5 g1f3 b8c6 f1c4 e7e6 e1g1 g8f6
info depth 11 seldepth 13 multipv 1 score cp 32 nodes 29040 nps 537777 hashfull 14 tbhits 0 time 54 pv e2e4 c7c5 c2c3 g8f6 e4e5 f6d5 d2d4
info depth 12 seldepth 14 multipv 1 score cp 38 nodes 41207 nps 564479 hashfull 18 tbhits 0 time 73 pv e2e4 e7e6 d2d4 d7d5 b1c3 d5e4 c3e4
info depth 13 seldepth 15 multipv 1 score cp 31 nodes 60308 nps 579884 hashfull 25 tbhits 0 time 104 pv e2e4 c7c5 g1f3 d7d6 d2d4 c5d4 f3d4 b8c6 c1e3 g8f6 b1c3 g7g6
info depth 14 seldepth 18 multipv 1 score cp 35 nodes 97789 nps 578633 hashfull 44 tbhits 0 time 169 pv d2d4 g8f6 c2c4 e7e6 g1f3 d7d5 g2g3 f8e7 b1c3 c7c5
info depth 15 seldepth 21 multipv 1 score cp 31 nodes 142447 nps 560814 hashfull 64 tbhits 0 time 254 pv d2d4 g8f6 c2c4 e7e6 g1f3 d7d5 c1g5 f8e7 b1c3 e8g8 e2e3 h7h6 g5h4 f6e4 h4e7 d8e7
info depth 16 seldepth 21 multipv 1 score cp 40 nodes 199887 nps 552174 hashfull 88 tbhits 0 time 362 pv d2d4 d7d5 c2c4 d5c4 g1f3 a7a6 b1c3 c7c6 a2a4
info depth 17 seldepth 19 multipv 1 score cp 32 nodes 268004 nps 552585 hashfull 120 tbhits 0 time 485 pv d2d4 g8f6 c2c4 e7e6 g2g3 d7d5 g1f3 f8e7 f1g2 e8g8 e1g1 c7c5 c4d5 e6d5 b1c3 h7h6 d4c5 e7c5
info depth 18 seldepth 19 multipv 1 score cp 32 nodes 277075 nps 553043 hashfull 122 tbhits 0 time 501 pv d2d4 g8f6 c2c4 e7e6 g2g3 d7d5 g1f3 f8e7 f1g2 e8g8 e1g1 c7c5 c4d5 e6d5 b1c3 h7h6 d4c5 e7c5
bestmove d2d4 ponder g8f6
```
So, movetime 150 seems to be sufficient to get some first insights in general. I would need help to understand, why it should be so fundamentally different in these rare cases. 

Ps: Even if I now use movetime 10, I get a result that is uci-conform:

```
ucinewgame
go movetime 10
info string NNUE evaluation using nn-ad9b42354671.nnue enabled
info depth 1 seldepth 0 multipv 1 score cp 0 nodes 1 nps 50 hashfull 0 tbhits 0 time 20 pv a2a3
bestmove a2a3
```

Pss: Even at movetime 1, there is some regular result, as I would need it:
```
ucinewgame
go movetime 1
info string NNUE evaluation using nn-ad9b42354671.nnue enabled
info depth 1 seldepth 0 multipv 1 score cp 0 nodes 1 nps 333 hashfull 0 tbhits 0 time 3 pv a2a3
bestmove a2a3
```

**Reply by bftjoe** [reply]
a2a3 is move::none or some other special value which means it did not complete a single depth. It is not surprising that it can't when you only give it 150 milliseconds. There is nothing to diagnose or solve here. Use a longer TC, or accept the results of your own foolishness.

**Reply by PGG106** [reply]
it's probably just the first move searched, but even with this incomplete search it produced an "info depth bla bla" report, so there really isn't a good reason as to why it doesn't at times

**Reply by PGG106** [reply]
while the TC is particularly aggressive and they are using an old sf version, there's no need to be as dismissive as you are being, the behaviour is definitely a tad weird

**Reply by PGG106** [reply]
well there's no use arguing about this without a fen to try to reproduce the issue or someone intimately familiar with sf 15.1 codebase

**Reply by Mercutio1243** [reply]
Thank you for your comments, and yes, I agree with a fen this will be more productive. I will report back, once I have identified a fen to investigate this further!

**Reply by bftjoe** [reply]
This type of output simply means no depth was every completed, and SF was unable to extract a ponder move from the transposition table, or else it would include a 2nd move, minimum. There is not some big mystery or anything worthy of looking into here.

**Reply by PGG106** [reply]
It can produce an info output after searching 1 node and can't in specific cases, how many nodes did it search in 150ms? How can they be less than 1? I see what you are saying but I don't see it matching the outputs we have

**Reply by bftjoe** [reply]
There is no such thing as searching 1 node in sf. It checks time every 512 nodes or at depth increment if it is a depth limit. It also checks time for the first call to search as callsCnt is initialized to 0. It is clear that sf simply doesn't think there is enough time to search even a single depth in rare cases with such a uselessly short time control. Otherwise, you get that result of the 512 node search.

**Reply by PGG106** [reply]
`info depth 1 seldepth 0 multipv 1 score cp 0 nodes 1 nps 333 hashfull 0 tbhits 0 time 3 pv a2a3` wouldn't this list 512 nodes then?, even if it was an incomplete depth

**Reply by PGG106** [reply]
i simply don't understand how an output like this `info depth 1 seldepth 0 multipv 1 score cp 0 nodes 1 nps 333 hashfull 0 tbhits 0 time 3 pv a2a3`, of a partial search, with just one node search reported, and the lack of the info message at all can cohexist.

**Reply by bftjoe** [reply]
As I stated already, callsCnt is initialized to 0, which forces a time check for the first node and means, in some rare cases, it can decide there is not enough time to search after the first node. 

If it decides that, then all the other output statements are skipped, and it just shows the first legal move.

**Reply by PGG106** [reply]
do you happen to know where the line of code that makes so that only best move is printed is?, i have looked around at the pv related stuff but i wasn't able to find it

**Reply by bftjoe** [reply]
The other printing happens inside the search loop, but this does not, so it always happens.

sync_cout << "bestmove " << UCI::move(bestThread->rootMoves[0].pv[0], rootPos.is_chess960());

https://github.com/official-stockfish/Stockfish/blob/632f1c21cd271e7c4c242fdafa328a55ec63b9cb/src/search.cpp#L202

**Reply by PGG106** [reply]
right, i should've caught that, specially because i do something similar in alex lol, i guess this is indeed intended behaviour, an "easier" way to fix this without having to radically change the code would be to move to a fixed nodes approach, estimating more or less how many nodes you want to search for in each position. afaik that should get around any time constraint shenanigans

**Reply by bftjoe** [reply]
Well, you could initialize callsCnt to a sensible value like 512 instead of 0 so there would always be decent output at the cost of possible timeouts. But it really is user error to set time this low.

**Reply by Mercutio1243** [reply]
I can only reiterate that the 150 ms movetime works splendidly in the 99.9999% of non-error cases, where I get a valid first impression on the position with a depths of 18-23, fully sufficient for my use case. But from the discussion, I am not sure that you are not open to my experience here. 

Apart from my personal view, as mentioned above the UCI protocol explicitly requires an info line before final bestmove statement without a minimum move time, ensuring that parsing the output works in standardized way.

So, if there is a code fix that assures output conformity, this might be a plus for everyone involved, independent of the individual move time preference?

**Reply by bftjoe** [reply]
There is nothing valid about setting such a low time, regardless of whether it works most of the time or not. You are abusing the engine. It is unfortunate that you continue to disregard people that actually understand the code.

Why would someone need to be open to your stubbornness to configure the engine properly?

There could be many things going on here, such as other loads on the machine taking up most of the 150 milliseconds you assigned to analyze each position.

**Reply by Disservin** [reply]
I don’t see how setting 150ms movetime is abusing the engine. Stockfish should report a final last uci line, if it doesn’t it’s not optimal

**Reply by bftjoe** [reply]
The easiest fix would be to initialize callsCnt to 512 or any other bigger number than 0.

**Reply by bftjoe** [reply]
Also, Stockfish violates UCI spec many ways, such as not accepting upper case in bool option values, requiring searchmoves to be at end of line, etc. This is hardly the first or last violation of the spec.

Technically, every engine that has file options and runs on Linux violates the UCI spec because it requires case insensitive paths and file names, which is not really possible on Linux because they are case sensitive.

I am looking forward to you submitting many 150ms per position tests to fishtest since this is not an abuse of the engine.
